File: extension/backend/content_filter.py
#!/usr/bin/env python3
"""
Content Filter Module for Socio.io
This module provides content moderation functionality for text and images.
"""

import re
import base64
import hashlib
import logging
import urllib.request
from urllib.parse import urlparse
import ssl
from cryptography.fernet import Fernet

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger('content-filter')

class ContentFilter:
    """Content moderation filter for text and images."""
    
    def __init__(self):
        """Initialize the content filter."""
        self.stats = {
            'text_filtered': 0,
            'images_filtered': 0,
            'total_requests': 0
        }
        
        # Initialize encryption key
        self.key = Fernet.generate_key()
        self.cipher = Fernet(self.key)
        
        # Define inappropriate content patterns - expanded for better coverage
        self.inappropriate_patterns = [
            r'\b(hate|violence|abuse|explicit|obscene)\b',
            r'\b(racist|sexist|discriminat(e|ion|ory))\b',
            r'\b(nsfw|porn|xxx|adult\s+content)\b',
            r'\b(kill|murder|assault|attack|threat)\b',
            r'\b(offensive|vulgar|profanity|swear)\b',
            r'\b(naked|nude|sex|sexual|erotic)\b'
        ]
        
        # Compile patterns for efficiency
        self.compiled_patterns = [re.compile(pattern, re.IGNORECASE) for pattern in self.inappropriate_patterns]
        
        logger.info("Content filter initialized")
    
    def filter_text(self, text):
        """
        Filter text content for inappropriate content.
        
        Args:
            text (str): The text to filter
            
        Returns:
            dict: Result of the filtering operation
        """
        self.stats['total_requests'] += 1
        
        if not text or len(text) < 3:
            return {
                'filtered': False,
                'reason': 'Text too short',
                'original': text,
                'modified': text
            }
        
        # Check for inappropriate content
        inappropriate = False
        matched_patterns = []
        
        for i, pattern in enumerate(self.compiled_patterns):
            if pattern.search(text):
                inappropriate = True
                matched_patterns.append(self.inappropriate_patterns[i])
        
        if inappropriate:
            # Encrypt the original content
            encrypted = self.encrypt_content(text)
            
            # Replace inappropriate content with asterisks
            modified_text = text
            for pattern in self.compiled_patterns:
                modified_text = pattern.sub(lambda m: '*' * len(m.group(0)), modified_text)
            
            self.stats['text_filtered'] += 1
            
            return {
                'filtered': True,
                'reason': 'Inappropriate content detected',
                'patterns': matched_patterns,
                'original': text,
                'modified': modified_text,
                'encrypted': encrypted
            }
        
        return {
            'filtered': False,
            'reason': 'No inappropriate content detected',
            'original': text,
            'modified': text
        }
    
    def filter_image(self, image_url):
        """
        Filter image content for inappropriate content.
        
        Args:
            image_url (str): URL of the image to filter
            
        Returns:
            dict: Result of the filtering operation
        """
        self.stats['total_requests'] += 1
        
        # Validate URL
        try:
            parsed_url = urlparse(image_url)
            if not parsed_url.scheme or not parsed_url.netloc:
                return {
                    'filtered': False,
                    'reason': 'Invalid URL',
                    'original': image_url,
                    'modified': image_url
                }
        except Exception as e:
            logger.error(f"Error parsing URL: {str(e)}")
            return {
                'filtered': False,
                'reason': f'Error parsing URL: {str(e)}',
                'original': image_url,
                'modified': image_url
            }
        
        # For demonstration purposes, we'll filter images based on URL patterns
        # In a real implementation, you would use image recognition APIs
        inappropriate_url_patterns = [
            r'nsfw',
            r'adult',
            r'xxx',
            r'porn',
            r'explicit'
        ]
        
        for pattern in inappropriate_url_patterns:
            if re.search(pattern, image_url, re.IGNORECASE):
                self.stats['images_filtered'] += 1
                
                # In a real implementation, you would replace with a placeholder image
                placeholder_image = "https://via.placeholder.com/400x300?text=Content+Filtered"
                
                return {
                    'filtered': True,
                    'reason': 'Potentially inappropriate image',
                    'original': image_url,
                    'modified': placeholder_image,
                    'encrypted': self.encrypt_content(image_url)
                }
        
        # For demonstration purposes, filter images more aggressively
        # This simulates an AI model making decisions
        import random
        if random.random() < 0.4:  # 40% chance of filtering - increased for better demo
            self.stats['images_filtered'] += 1
            
            # In a real implementation, you would replace with a placeholder image
            placeholder_image = "https://via.placeholder.com/400x300?text=Content+Filtered"
            
            return {
                'filtered': True,
                'reason': 'Potentially inappropriate content detected',
                'original': image_url,
                'modified': placeholder_image,
                'encrypted': self.encrypt_content(image_url)
            }
        
        return {
            'filtered': False,
            'reason': 'No inappropriate content detected',
            'original': image_url,
            'modified': image_url
        }
    
    def encrypt_content(self, content):
        """
        Encrypt content for secure storage.
        
        Args:
            content (str): Content to encrypt
            
        Returns:
            str: Encrypted content as a base64 string
        """
        try:
            # Convert string to bytes
            content_bytes = content.encode('utf-8')
            
            # Encrypt the content
            encrypted_bytes = self.cipher.encrypt(content_bytes)
            
            # Convert to base64 string for storage
            encrypted_str = base64.b64encode(encrypted_bytes).decode('utf-8')
            
            return encrypted_str
        except Exception as e:
            logger.error(f"Error encrypting content: {str(e)}")
            return ""
    
    def decrypt_content(self, encrypted):
        """
        Decrypt previously encrypted content.
        
        Args:
            encrypted (str): Encrypted content as a base64 string
            
        Returns:
            str: Decrypted content
        """
        try:
            # Convert base64 string to bytes
            encrypted_bytes = base64.b64decode(encrypted)
            
            # Decrypt the content
            decrypted_bytes = self.cipher.decrypt(encrypted_bytes)
            
            # Convert bytes to string
            decrypted_str = decrypted_bytes.decode('utf-8')
            
            return decrypted_str
        except Exception as e:
            logger.error(f"Error decrypting content: {str(e)}")
            return ""
    
    def get_stats(self):
        """
        Get statistics about the content filter.
        
        Returns:
            dict: Statistics about filtered content
        """
        return self.stats

File: backend/content_filter.js
/**
 * Content Filter Module
 * Provides content moderation for text and images
 */

const fs = require('fs').promises;
const path = require('path');
const crypto = require('crypto');
const dotenv = require('dotenv');
const winston = require('winston');
const { VertexAI } = require('@google-cloud/vertexai');

// Load environment variables
dotenv.config();

// Configure logging
const logger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.printf(({ timestamp, level, message }) => {
      return `${timestamp} - ${level}: ${message}`;
    })
  ),
  transports: [
    new winston.transports.Console()
  ]
});

class ContentFilter {
  constructor() {
    // Initialize stats
    this.stats = {
      text_filtered: 0,
      images_filtered: 0,
      total_requests: 0,
      python_available: false,
      google_cloud_available: false,
      vertex_ai_available: false
    };

    // Initialize Vertex AI
    try {
      const projectId = process.env.GOOGLE_CLOUD_PROJECT;
      const location = process.env.VERTEX_AI_LOCATION || 'us-central1';
      const modelName = process.env.VERTEX_AI_MODEL || 'gemini-1.5-flash-001';
      
      if (!projectId) {
        throw new Error('GOOGLE_CLOUD_PROJECT environment variable is not set');
      }
      
      logger.info(`Initializing Vertex AI with project: ${projectId}, location: ${location}, model: ${modelName}`);
      
      this.vertexAi = new VertexAI({
        project: projectId,
        location: location,
      });
      
      this.generativeModel = this.vertexAi.getGenerativeModel({
        model: modelName,
        generationConfig: {
          temperature: 0.1,
          topK: 40,
          topP: 0.8,
          maxOutputTokens: 1024,
        }
      });
      
      this.stats.vertex_ai_available = true;
      this.stats.google_cloud_available = true;
      logger.info('Vertex AI initialized successfully');
    } catch (error) {
      logger.error(`Error initializing Vertex AI: ${error.message}`);
      this.vertexAi = null;
      this.generativeModel = null;
    }

    // Check if Python bridge is available
    try {
      const pythonBridge = require('./python_bridge');
      this.pythonBridge = pythonBridge;
      this.stats.python_available = true;
      logger.info('Python bridge initialized successfully');
    } catch (error) {
      logger.error(`Error initializing Python bridge: ${error.message}`);
      this.pythonBridge = null;
    }

    logger.info('Content filter initialized');
  }

  /**
   * Filter text content for inappropriate content
   * @param {string} text - The text to filter
   * @returns {Object} - Filtering results
   */
  async filterText(text) {
    try {
      this.stats.total_requests++;
      
      // Use Vertex AI if available
      if (this.generativeModel) {
        try {
          logger.info('Using Vertex AI for text analysis');
          
          const prompt = `
          Analyze the following text and identify if it contains inappropriate content such as:
          - Hate speech
          - Profanity
          - Violence
          - Sexual content
          - Discrimination
          - Personal attacks
          
          Return a JSON object with these fields:
          - "filtered": boolean (true if inappropriate content is detected)
          - "reason": string (explanation of why content was filtered or not)
          - "modified": string (the original text with inappropriate words replaced by asterisks)
          
          Text to analyze: "${text}"
          
          Respond with ONLY the JSON object. No other text.
          `;
          
          const request = {
            contents: [{ role: "user", parts: [{ text: prompt }] }]
          };
          
          const response = await this.generativeModel.generateContent(request);
          const responseText = response.response.candidates[0].content.parts[0].text;
          
          // Try to parse the JSON response
          try {
            // Clean up the response text to ensure it's valid JSON
            const cleanedResponse = responseText.replace(/```json|```/g, '').trim();
            const result = JSON.parse(cleanedResponse);
            
            if (result.filtered) {
              this.stats.text_filtered++;
            }
            
            return {
              filtered: result.filtered,
              reason: result.reason,
              original: text,
              modified: result.modified || text,
              encrypted: this.encryptContent(text),
              analysis: {
                provider: 'vertex_ai',
                model: process.env.VERTEX_AI_MODEL || 'gemini-1.5-flash-001'
              }
            };
          } catch (parseError) {
            logger.error(`Error parsing Vertex AI response: ${parseError.message}`);
            logger.error(`Raw response: ${responseText}`);
            throw new Error('Failed to parse Vertex AI response');
          }
        } catch (vertexError) {
          logger.error(`Vertex AI analysis error: ${vertexError.message}`);
          // Fall back to regex-based filtering
        }
      }
      
      // Simple regex-based filtering as fallback
      logger.warn('Using fallback text filtering (regex-based)');
      
      // Simple regex-based filtering
      const inappropriate = /hate|violence|abuse|explicit|obscene|racist|sexist|discriminat|nsfw|porn|xxx/i.test(text);
      
      if (inappropriate) {
        // Replace inappropriate content with asterisks
        const modifiedText = text.replace(/hate|violence|abuse|explicit|obscene|racist|sexist|discriminat|nsfw|porn|xxx/gi, 
          match => '*'.repeat(match.length));
        
        this.stats.text_filtered++;
        
        return {
          filtered: true,
          reason: 'Inappropriate content detected',
          original: text,
          modified: modifiedText,
          encrypted: this.encryptContent(text)
        };
      }
      
      return {
        filtered: false,
        reason: 'No inappropriate content detected',
        original: text,
        modified: text
      };
    } catch (error) {
      logger.error(`Error filtering text: ${error.message}`);
      
      // Return a safe fallback response
      return {
        filtered: false,
        reason: 'Error during content analysis',
        original: text,
        modified: text,
        error: error.message
      };
    }
  }

  /**
   * Filter image content for inappropriate content
   * @param {string} url - URL of the image to filter
   * @returns {Object} - Filtering results
   */
  async filterImage(url) {
    try {
      this.stats.total_requests++;
      
      // Use Vertex AI if available
      if (this.generativeModel) {
        try {
          logger.info('Using Vertex AI for image URL analysis');
          
          const prompt = `
          Analyze this image URL and determine if it might contain inappropriate content:
          ${url}
          
          Based only on the URL (not the actual image content), check for keywords or patterns that suggest:
          - Adult content
          - Violence
          - Hate speech
          - Explicit material
          
          Return a JSON object with these fields:
          - "filtered": boolean (true if URL suggests inappropriate content)
          - "reason": string (explanation of why URL was flagged or not)
          - "confidence": number (0-1 indicating confidence in the assessment)
          
          Respond with ONLY the JSON object. No other text.
          `;
          
          const request = {
            contents: [{ role: "user", parts: [{ text: prompt }] }]
          };
          
          const response = await this.generativeModel.generateContent(request);
          const responseText = response.response.candidates[0].content.parts[0].text;
          
          // Try to parse the JSON response
          try {
            // Clean up the response text to ensure it's valid JSON
            const cleanedResponse = responseText.replace(/```json|```/g, '').trim();
            const result = JSON.parse(cleanedResponse);
            
            if (result.filtered) {
              this.stats.images_filtered++;
              
              return {
                filtered: true,
                reason: result.reason,
                original: url,
                modified: 'https://via.placeholder.com/400x300?text=Content+Filtered',
                encrypted: this.encryptContent(url),
                analysis: {
                  provider: 'vertex_ai',
                  model: process.env.VERTEX_AI_MODEL || 'gemini-1.5-flash-001',
                  confidence: result.confidence
                }
              };
            }
            
            return {
              filtered: false,
              reason: result.reason,
              original: url,
              modified: url,
              analysis: {
                provider: 'vertex_ai',
                model: process.env.VERTEX_AI_MODEL || 'gemini-1.5-flash-001',
                confidence: result.confidence
              }
            };
          } catch (parseError) {
            logger.error(`Error parsing Vertex AI response: ${parseError.message}`);
            logger.error(`Raw response: ${responseText}`);
            throw new Error('Failed to parse Vertex AI response');
          }
        } catch (vertexError) {
          logger.error(`Vertex AI analysis error: ${vertexError.message}`);
          // Fall back to regex-based filtering
        }
      }
      
      // Simple URL-based filtering as fallback
      logger.warn('Using fallback image filtering (URL-based)');
      
      // Simple URL-based filtering
      const inappropriate = /nsfw|adult|xxx|porn|explicit/i.test(url);
      
      if (inappropriate) {
        this.stats.images_filtered++;
        
        return {
          filtered: true,
          reason: 'Potentially inappropriate image URL',
          original: url,
          modified: 'https://via.placeholder.com/400x300?text=Content+Filtered',
          encrypted: this.encryptContent(url)
        };
      }
      
      return {
        filtered: false,
        reason: 'No inappropriate content detected in URL',
        original: url,
        modified: url
      };
    } catch (error) {
      logger.error(`Error filtering image: ${error.message}`);
      
      // Return a safe fallback response
      return {
        filtered: false,
        reason: 'Error during content analysis',
        original: url,
        modified: url,
        error: error.message
      };
    }
  }

  /**
   * Encrypt content for secure storage
   * @param {string} content - Content to encrypt
   * @returns {string} - Encrypted content (base64)
   */
  encryptContent(content) {
    try {
      // Simple base64 encoding for the minimal version
      return Buffer.from(content).toString('base64');
    } catch (error) {
      logger.error(`Error encrypting content: ${error.message}`);
      return '';
    }
  }

  /**
   * Decrypt previously filtered content
   * @param {string} encrypted - Encrypted content
   * @returns {string} - Decrypted content
   */
  async decryptContent(encrypted) {
    try {
      // Simple base64 decoding for the minimal version
      return Buffer.from(encrypted, 'base64').toString('utf-8');
    } catch (error) {
      logger.error(`Error decrypting content: ${error.message}`);
      throw new Error('Failed to decrypt content');
    }
  }

  /**
   * Get statistics about content filtering
   * @returns {Object} - Statistics
   */
  getStats() {
    return {
      text_filtered: this.stats.text_filtered,
      images_filtered: this.stats.images_filtered,
      total_requests: this.stats.total_requests,
      python_available: this.stats.python_available,
      google_cloud_available: this.stats.google_cloud_available,
      vertex_ai_available: this.stats.vertex_ai_available
    };
  }
}

module.exports = ContentFilter;